src/quickgui/qgsquickmapcanvasmap.h

```diff
@@ -0,0 +1,202 @@
+     */
+    Q_PROPERTY( bool incrementalRendering READ incrementalRendering WRITE setIncrementalRendering NOTIFY incrementalRenderingChanged )
+
+  public:
+    //! Create map canvas map
+    QgsQuickMapCanvasMap( QQuickItem *parent = nullptr );
+    ~QgsQuickMapCanvasMap() = default;
+
+    virtual QSGNode *updatePaintNode( QSGNode *oldNode, QQuickItem::UpdatePaintNodeData * ) override;
+
+    //! \copydoc QgsQuickMapCanvasMap::mapSettings
+    QgsQuickMapSettings *mapSettings() const;
+
+    //! \copydoc QgsQuickMapCanvasMap::freeze
+    bool freeze() const;
+
+    //! \copydoc QgsQuickMapCanvasMap::freeze
+    void setFreeze( bool freeze );
+
+    //! \copydoc QgsQuickMapCanvasMap::isRendering
+    bool isRendering() const;
+
+    //! \copydoc QgsQuickMapCanvasMap::mapUpdateInterval
+    int mapUpdateInterval() const;
+
+    //! \copydoc QgsQuickMapCanvasMap::mapUpdateInterval
+    void setMapUpdateInterval( int mapUpdateInterval );
+
+    //! \copydoc QgsQuickMapCanvasMap::incrementalRendering
+    bool incrementalRendering() const;
+
+    //! \copydoc QgsQuickMapCanvasMap::incrementalRendering
+    void setIncrementalRendering( bool incrementalRendering );
+
+  signals:
+
+    /**
+     * Signal is emitted when a rendering is starting
+     */
+    void renderStarting();
+
+    /**
+     * Signal is emitted when a canvas is refreshed
+     */
+    void mapCanvasRefreshed();
+
+    //! \copydoc QgsQuickMapCanvasMap::freeze
+    void freezeChanged();
+
+    //! \copydoc QgsQuickMapCanvasMap::isRendering
+    void isRenderingChanged();
+
+    //!\copydoc QgsQuickMapCanvasMap::mapUpdateInterval
+    void mapUpdateIntervalChanged();
+
+    //!\copydoc QgsQuickMapCanvasMap::incrementalRendering
+    void incrementalRenderingChanged();
+
+  protected:
+    virtual void geometryChanged( const QRectF &newGeometry, const QRectF &oldGeometry ) override;
+
+  public slots:
+    //! Stop map rendering
+    void stopRendering();
+
+    /**
+     * Set map setting's extent (zoom the map) on the center by given scale
+     */
+    void zoom( QPointF center, qreal scale );
+
+    /**
+     * Set map setting's extent (pan the map) based on the difference of positions
+     */
+    void pan( QPointF oldPos, QPointF newPos );
+
+    /**
+     * Refresh the map canvas.
+     * Does nothing when output size of map settings is not set
+     */
+    void refresh();
+
+  private slots:
+    void refreshMap();
+    void renderJobUpdated();
+    void renderJobFinished();
+    void onWindowChanged( QQuickWindow *window );
+    void onScreenChanged( QScreen *screen );
+    void onExtentChanged();
+    void onLayersChanged();
+
+  private:
+
+    /**
+     * Should only be called by stopRendering()!
+     */
+    void destroyJob( QgsMapRendererJob *job );
+    QgsMapSettings prepareMapSettings() const;
+    void updateTransform();
+    void zoomToFullExtent();
+
+    QgsQuickMapSettings *mMapSettings;
```

Comment:
Where is this deleted?

Reply:
never, added destructor

Reply:
unique_ptr?

Reply:
`QgsQuickMapSettings` is QObject, I think it should not be used in combination with `unique_ptr`

Reply:
I disagree - a parent less QObject can still be a unique ptr, and it's really no different then manually deleting in the destructor like you've done here. But it adds the memory safety which comes with unique_ptr.

Reply:
Even a QObject with parent can be a unique_ptr, it will be removed from its parent's children list on deletion. It just doesn't really make sense too often and is rather confusing, so different approaches are (almost?) always preferable.

Reply:
Actually that's a bit dangerous - it's possible that a parented QObject is deleted before the unique_ptr deleter, causing a crash. (This has been seen in QGIS code).

Reply:
Good point, silly me.

Reply:
replaced with unique_ptr